CLUSTER.md

```diff
@@ -0,0 +1,450 @@
+
+![Flotilla High Level overview](images/high_level.jpg?raw=true "High Level Overview")
+
+### The Topology Director
+
+The Topology Director serves as the source of truth for the cluster topology (excluding primary/replica status) and helps propagate
+information from Failover Coordinators. It never handles any user data, only topology metadata.
+Since it owns the metadata of the entire cluster it can allow changing this metadata while preserving the strong
+consistency guarantees we desire.
+For example - it can ensure that if a Shard holds a slot marked as Migrating, there's a corresponding Shard with the slot marked as Importing.
+We also rely on the Failover Coordinator to distribute the primary/replica state of individual nodes - This state is received periodically
+from every Failover Coordinator, cached in the Topology Director and relayed to all other Failover Coordinators when they send their state.
+
+- A completely reactive entity
+- Receives requests to alter the cluster topology and adjusts the internal state as needed
+- Validates change requests for correctness
+- Responds to topology queries by the other components
+
+#### Interfaces
+
+The Topology Director interface is fully declarative other than join/remove flows. This allows for easy orchestration via various tools.
+
+With the administrator, directly for creating the cluster / updating and removing nodes:
+
+**FLOTILLA.TOPOLOGY**: Similar to CLUSTER.SLOTS/NODES - response lists all the Flotilla nodes and their metadata  
+**FLOTILLA.REMOVE_NODE**: Remove a given node from the cluster  
+**FLOTILLA.ASSIGN_SLOTS**: Assign a set slots to specified Shard  
+**FLOTILLA.UPDATE_NODES**: Change the topology of a list of nodes    
+
+    Update/Assign_slots commands can specify the epoch they are based on in order to allow flotilla to reject outdated requests.
+    Update allows:
+    - Taking ownership of unassigned slots
+    - Moving an unallocated node / shard between Failover Coordinators
+  
+With the administrator via the individual node for adding nodes / Failover Coordinators to Flotilla:
+
+**FLOTILLA.CREATE_TD**: Create the Flotilla Cluster from the current node  
+**FLOTILLA.CREATE_FC**: Create a new Failover Coordinator entity from the current node  
+**FLOTILLA.JOIN_TD**: Add the node to be part of the Topology Director quorum  
+**FLOTILLA.JOIN_FC**: Add the node to be part of a given Failover Coordinator  
+**FLOTILLA.JOIN_NODE**: Add the data node to Flotilla  
+    
+    All Join/Creation requests can specify a desired Shard ID/slot range up front or set it later using an UPDATE command.
+    Responses contain the new node/Failover Coordinator IDs and the topology information needed for the new node to start functioning - a list
+    of all Topology Director nodes and all the nodes belongining to the relevant Failover Coordinator.  
```

Comment:
I'm pretty sure this boils down to the "Push vs Pull" debate, and we can think of equivalent pros/cons for both sides.
Need to think about it some more.

Will say that if we switch - we should do it across all layers (So the TD would connect to the nodes to join and to the FC to get their state in addition to the FC connecting to the data nodes).

Reply:
I agree we need to think about it more. Intuitively I find it easier to think in terms of pull when a a large fleet of N worker entities talks to a single control plane element.

Some arguments:
* As @madolson mentioned, we need that mode anyway for registration.
* A node that can't talk/doesn't hear from the CP has more information about why that's the case (i.e. connection error, auth error, etc. vs just sitting there waiting for the call).

Reply:
Took my time to think about this one.
To make sure we're aligned terminology wise  -- Push == Node initiates connections to FC, FC to TD (like current spec).

I'd argue that the only true advantage of Pull here is the ability to isolate the control plane network from the data plane. All other Pros have an equivalent Con. Considering the original points raised:
1. As mentioned, discovery becomes significantly more complex. If we want to prevent IP changes from breaking the cluster we now need to maintain internal DNS resolution for all data nodes and not just for the FC/TD (eg - thousands of records instead of <10 in most cases). 
2.  Moving this logic to the FC layer makes it more complicated and increases the impact it will have in the naive deployment cases where it runs side-by-side on a data node (and also means that alternative implementations become more difficult). It also means that a data node must wait for the FC to connect to it in order to initiate failover - instead of immediately triggering one when it notes it is unable to connect to the primary/receives the failover command.
3. Agree about the network isolation aspect, though it is worth mentioning this is only relevant for folks running with completely separate data/FC/TD nodes AND  (probably only the bigger Redis-as-a-Service vendors).
4. A misbehaving FC will generate load on plane components, implementing this logic on the FC side is more difficult then simple jitter+backoff and might make it harder to predict failover times, means that implementing a TD/FC becomes more complicated.
5. FCs will keep retrying dead/unreachable nodes forever.
6. I'm not sure why switching to Pull makes the API simpler. It makes some of the implementation easier / harder and moves some of the more networking oriented bits into the FC, but unless you're proposing a bigger change it does not remove the need for the interface itself.
7. Is that really significant? As you wrote we still need to persist the identity + some cluster ID/secret so there's still the need for storage.

In addition to the above (and the other cons listed in the thread):
- Joining nodes become a two step operation for 'simple' admins (Will have to go to the data node to configure the cluster secret, and then send the request to the TD to add the new node).
- Without significant other changes, Heartbeats/Failovers will become a two/four-step operation (instead of two/three): FC asks for HB, Receives a response with an out-of-date epoch, sends the relevant topology, receives the updated HB.

Reply:
@yossigo / @madolson / @PingXie - thoughts? If there's no consensus we should probably schedule some online session to discuss this at depth.

Reply:
@ushachar Aren't we going to have a separate session any ways to go over this? I'm re-reading through it now and taking some notes. I was going to post more of a `summary` of my thoughts tomorrow. The document is more clear and concise now, and I at least have a better handle of the various trade offs, and would feel pretty comfortable just getting together to commit to something.

Reply:
> As mentioned, discovery becomes significantly more complex. If we want to prevent IP changes from breaking the cluster we now need to maintain internal DNS resolution for all data nodes and not just for the FC/TD (eg - thousands of records instead of <10 in most cases).

This might be a fair point highlighting my cloud bias :) The folks I've talked to from other places say that this isn't a big deal, as in most cases nodes are still typically either assigned a permanent IP or hostname within most production networks. 

> FCs will keep retrying dead/unreachable nodes forever.

This gets at whether or not a cluster is an interchangeable bag of nodes, or nodes that represent a specific topology. The trend seems to the latter, so I assume retrying forever is the correct behavior, until it's reconfigured.

> Is that really significant? As you wrote we still need to persist the identity + some cluster ID/secret so there's still the need for storage.

These items are persistent and don't really change over time (presumably), which makes it easier to persist them. They could be provided on an ephemeral disk or startup arguments. So, yes, it is somewhat significant. 

> I'm not sure why switching to Pull makes the API simpler. It makes some of the implementation easier / harder and moves some of the more networking oriented bits into the FC, but unless you're proposing a bigger change it does not remove the need for the interface itself.

The specific point is that I see it as very useful if data nodes only know about the FC interface, they don't know about the TD. I see this as useful because I view the TD as ultimately an optimization for specialized configurations, as a lot of other folks have addressed, and don't want to mandate it as part of the interface. I don't see how this doesn't dramatically simplify contract and flows. I can see plenty of implementations, including a very simplified version of the one being proposed, which don't have the specific forced architecture of a separate TD and FC. 

Another sub point which I think was missed, is that the FC can reach out to the node, but the node can still follow up by pushing heart beats back at any time. These should be long-lived established connections, so there is no need for the establisher of the connection to necessarily drive all of the communication.

Reply:
@madolson I'm afraid I still don't understand how switching to a Pull based approach makes the API simpler. Unless you're thinking about a much bigger change then just Push->Pull then beyond JOIN vs ADD the change is just in who opens the relevant connection, not in how the flow works

If the main issue is around the Node->TD connection for the initial JOIN request we can always support proxing that through the relevant FC... The node doesn't really care where the upstream request is really handled.

>This might be a fair point highlighting my cloud bias :) The folks I've talked to from other places say that this isn't a big deal, as in most cases nodes are still typically either assigned a permanent IP or hostname within most production networks.

This is often (but not always) true when you're talking VMs / physical instances, but is very much false when in a containerized environment where the IPs change whenever a pod is rescheduled.

> Another sub point which I think was missed, is that the FC can reach out to the node, but the node can still follow up by pushing heart beats back at any time. These should be long-lived established connections, so there is no need for the establisher of the connection to necessarily drive all of the communication.

This is similar to long polling... it introduces more complexity on the FC side.

Reply:
> I'm afraid I still don't understand how switching to a Pull based approach makes the API simpler. Unless you're thinking about a much bigger change then just Push->Pull then beyond JOIN vs ADD the change is just in who opens the relevant connection, not in how the flow works

I didn't understand this statement. The change is not just about who opens the relevant connection, but about who is responsible for maintaining the connection. Typically the control plane is responsible for establishing health check connections and sending messages, not the data plane. In the current model, the data path is responsible for knowing where the control plane is and sending messages to it. I would expect a model where the data path knows nothing, and receives updates from the control plane about state.

Both models can reasonably work. 

>If the main issue is around the Node->TD connection for the initial JOIN request we can always support proxing that through the relevant FC... The node doesn't really care where the upstream request is really handled.

Sure, if we stick with the current approach, I think that also works.

> This is often (but not always) true when you're talking VMs / physical instances, but is very much false when in a containerized environment where the IPs change whenever a pod is rescheduled.

I don't know enough about containerized environments to make a claim about this. I've been told this is trivial to solve with hostnames, but I myself am not an active user of kubernetes or some similar system to know if that is true.

> This is similar to long polling... it introduces more complexity on the FC side.

I'm not convinced this is materially much more complex.

Reply:
I have a change of heart. I now view FCs/TDs metadata nodes of the cluster hence part of the data plane as opposed to the control plane. The difference between them matters to me because the clean separation of CP/DP is no longer a concern with this view. This view is also consistent with how we have been regularly discussing cluster v1 and v2 side by side and considering them different solutions for the same clustering problem.  After mulling over this some more, I now feel that the push model is easier to work with and yields more reliability. I see 3 scenarios being affected by the pull-vs-push decision: steady state with node failures, cluster bootstrapping, and scaling-out. The latter two involve the discovery problem that @ushachar called out.

1. steady state with node failures
I now think the push model is more reliable than the pull model as it involves less work on the FC side (the repetitive polling of down nodes). The thundering herd risk associated with the push model can be mitigated by a rate limiter on the FC side but the repetitive polling would add more stress on the FC node especially in the large scale event case and when a large cluster is involved. Given that FCs (and TDs) own the cluster metadata, their health is of higher priority IMO.
2. cluster bootstrapping and scaling-out
I agree discovery is at the core in these two scenarios. If my understanding is correct, the data node will be created/bootstrapped "out-of-band", meaning not through the new cluster V2 APIs. With the "pull" model, as has been pointed out, the identities of the new data nodes need to be noted down somewhere after their creation. We then need to partition the new data nodes and assign them to their corresponding FCs. This can also be achieved by passing the identities of the new data nodes to the TDs and then having the TDs split up the data nodes somehow. With the "push" model, FCs and TDs can be created first. Data nodes can then be created/bootstrapped with their owning FC's identity. In theory, it is also possible to bootstrap data nodes with just TDs' network identities and again have TDs redirect the new data nodes to the proper FCs. There seems to be less states to manage and again less work on the FCs/TDs side.

Reply:
After a discussion, the conclusion was that although both options were possible, since the current design revolves aren't reconciliation it would be more effort to change.